### PR TITLE
[serdes] iterative serialization

### PR DESCRIPTION
**Pull request by alangenfeld**
Change how we drive serialization from the two step "pack" (turn whole tree in to primitives) then serialize to one where we wrap object types in a fake dict and iteratively process them as json iterencode walks the tree. 

original prototype https://github.com/dagster-io/dagster/pull/21595

## How I Tested These Changes

these raw results https://gist.github.com/alangenfeld/6673e4a8718ccea2dc4b287f811939eb fed through an llm to make this table

Target | Size | serialize_value (mean) | pack_value (mean) | serialize_value (% change) | pack_value (% change)
-- | -- | -- | -- | -- | --
0 | 68.7KiB | 1.679 (master) / 1.569 (al) | 1.100 (master) / 1.020 (al) | -6.5% | -7.8%
1 | 9.7MiB | 688.566 (master) / 398.689 (al) | 631.490 (master) / 424.501 (al) | -42.4% | -33.4%
2 | 25.8MiB | 2366.767 (master) / 1855.869 (al) | 2082.907 (master) / 1898.141 (al) | -21.6% | -10.3%
3 | 9.1MiB | 282.533 (master) / 234.455 (al) | 212.199 (master) / 228.063 (al) | -17.2% | +6.9%
4 | 2.5MiB | 50.936 (master) / 49.849 (al) | 33.145 (master) / 34.069 (al) | -1.7% | +2.7%
5 | 635.7KiB | 14.285 (master) / 14.051 (al) | 9.332 (master) / 9.861 (al) | -1.7% | +5.3%
6 | 42.6KiB | 0.934 (master) / 0.997 (al) | 0.604 (master) / 0.649 (al) | +6.3% | +7.0%
7 | 40.6MiB | 1305.244 (master) / 927.526 (al) | 990.155 (master) / 1045.181 (al) | -28.7% | +5.3%

we see the most improvement on large payloads

memory profiling `Target 3`:  ` 173.451MB` -> `85.128MB` ~50% reduction 
https://gist.github.com/alangenfeld/8c67ec41729f99fe78c73d70bfe8d080

### CODE REVIEW

**Comment by alangenfeld** [comment]
* **#21607** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21607?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @alangenfeld and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_serdes/serdes.py [hunk 1]
```diff
@@ -883,18 +910,63 @@ def _pack_value(
...
+    def __init__(
+        self,
+        obj: SerializableObject,
+        whitelist_map: WhitelistMap,
+        descent_path: str,
+    ):
+        self._obj = obj
+        self._whitelist_map = whitelist_map
+        self._descent_path = descent_path
+        # populate backing native dict to work around c optimizations in json serializer
```
Comment (schrockn) [inline python_modules/dagster/dagster/_serdes/serdes.py:950]: I think this is worthy of a more in-depth explanation

File: python_modules/dagster/dagster/_serdes/serdes.py [hunk 2]
```diff
@@ -805,12 +816,18 @@ def pack_value(
         * frozenset
     """
     descent_path = _root(val) if descent_path is None else descent_path
-    return _pack_value(val, whitelist_map=whitelist_map, descent_path=descent_path)
+    return _prepare_value(
+        val,
```
Comment (schrockn) [inline python_modules/dagster/dagster/_serdes/serdes.py:820]: is `pack_value` above a public API? Is that we are not changing the name of it? It's a bit confusing at first plush not that it is not actually packing a value, but instead wrapping it in the wrapper thing in prep for iterative serailization.

**Review by schrockn — CHANGES_REQUESTED** [review]
Brilliant PR. I think with a little bikeshedding/renaming we could make things more clear. As written 'prepare' and 'pack' are intermixed in confusing ways.  And `_SerdesObjectWrapper` definitely smacks of a "couldn't think of anything so typed some words" name.

My mental model is that we

1) Have an in-memory object graph
2) The "prepare" step wraps it in some intermediate, temporary object structure that abides by an interface in order to make it amenable iteration by the json encoder.
3) We pass that intermediate structure to json.dump and it never has to build up the entire serialized representation into memory. It can stream to its intended target.

The entire purpose of `_SerdesObjectWrapper` is provide a json-dump-compatible streaming interface to our supported object types (namedtuple, BaseModel, and dataclasses). And they are created on demand as the traversal unfolds, I think?

Tl;dr Idea is sound and execution is elegant/brilliant but I think we can do better on names and docs so the next person who touches this or has to fix a bug can get up to speed more quickly.

**Comment by salazarm** [comment]
Could each object be a proxy that lazily deserializes the value? At least for now we have a lot of queries scoped to specific object types (eg. sensors) that end up deserializing the whole workspace.

File: python_modules/dagster/dagster/_serdes/serdes.py (hunk 2 above)
Comment (alangenfeld) [reply python_modules/dagster/dagster/_serdes/serdes.py:820]: Yea `pack_value` is a public API for cases when a structure of primitives is needed without going directly to json string. Often this is used to as an intermediate before serializing to json but nesting in some other structure. 

To support both `pack_value` and `serialize_value` , their shared traversal function `_prepare_value` takes an `object_handler` function to either go all the way to primitives in one shot [1] or just wrap for incremental serialization [2]

File: python_modules/dagster/dagster/_serdes/serdes.py [hunk 3]
```diff
@@ -883,18 +910,63 @@ def _pack_value(
...
 
 
+def _pack_object(
+    obj: SerializableObject, whitelist_map: WhitelistMap, descent_path: str
+) -> Mapping[str, JsonSerializableValue]:
+    # the object_handler for _prepare_value to produce dicts for objects
+
+    klass_name = obj.__class__.__name__
+    serializer = whitelist_map.get_object_serializer(klass_name)
+    return dict(serializer.pack_items(obj, whitelist_map, _pack_object, descent_path))
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_serdes/serdes.py:932]: [1]

File: python_modules/dagster/dagster/_serdes/serdes.py [hunk 4]
```diff
@@ -883,18 +910,63 @@ def _pack_value(
...
+        )
+
+
+def _wrap_object(
+    obj: SerializableObject,
+    whitelist_map: WhitelistMap,
+    descent_path: str,
+) -> "_SerdesObjectWrapper":
+    # the object_handler for _prepare_value to use in conjunction with json.dumps for iterative serialization
+    return _SerdesObjectWrapper(obj, whitelist_map, descent_path)
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_serdes/serdes.py:967]: [2]

**Review by alangenfeld — COMMENTED** [review]
> My mental model is ...

yep thats right, heres a visualization of before and after

Screenshot 2024-05-03 at 9 54 28 AM


> Could each object be a proxy that lazily deserializes the value? At least for now we have a lot of queries scoped to specific object types (eg. sensors) that end up deserializing the whole workspace.

An interesting idea that could be explored in another PR. I am not sure how to implement this one in a way that doesn't cascade in to having to update an enormous amount of code. 

> `_SerdesObjectWrapper`

Happy to take name ideas on this one,  `OnDemandSerializationWrapper` maybe?

File: python_modules/dagster/dagster/_serdes/serdes.py (hunk 3 above)
Comment (schrockn) [reply python_modules/dagster/dagster/_serdes/serdes.py:932]: Got it. Should we use the word "traverse" in the name of `_prepare_value`? I think that would clarify things quite a bit.

**Review by schrockn — APPROVED** [review]
Great. Nicely done.
